### PR TITLE
Feat/cli/appmanager/publish

### PR DESCRIPTION
**Pull request by eikeland**
## Why

New fusion-framework-cli commands for managing apps.

**Commands**

- [x] `build-publish` - create zip bundle and runs `upload` and `tag` on the bundle.
- [x] `build-upload` - uploads provided _`--bundle <zipbundle>`_
- [x] `build-tag` - tags provided _`--version <semver>`_ with 'preview' | 'latest'
- [x] `build-config` -  Will upload the generated config to the version provided.
- [x] `build-manifest` - Will generate manifest.

**TODO**
- [ ] Change ``UserSettingsClient`` from @equinor/fusion in portal.
- [x] Apps bundle should be loaded from apps-proxy
- [x] CLI: Add notations to AppManifestExport in fusion-framework-cli app manifest.
- [x] module-app: find a better way to handle changing appKey in ApplicationManifest when creating legacy app component in apploader.

Create prereleases of framework packages from this pr and install them in portal to test.

Reference [AB#52910](https://statoil-proview.visualstudio.com/3dadd756-da44-4bb3-8874-330932214dff/_workitems/edit/52910)

Reference [AB#52917](https://statoil-proview.visualstudio.com/3dadd756-da44-4bb3-8874-330932214dff/_workitems/edit/52917)

Reference [AB#52915](https://statoil-proview.visualstudio.com/3dadd756-da44-4bb3-8874-330932214dff/_workitems/edit/52915)

### Check off the following:
- [x] Confirm that I checked changes to branch which I am merging into.
  - _I have validated included files_
  - _My code does not generate new linting warnings_
  - _My PR is not a duplicate, [check existing pr`s](https://github.com/equinor/fusion-framework/pulls)_
- [x] Confirm that the I have completed the [self-review checklist](https://github.com/equinor/fusion-framework/blob/main/contributing/self-review.md).

- [x] Confirm that my changes meet our [code of conduct](https://github.com/equinor/fusion-framework/blob/main/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: packages/cli/src/bin/utils/app-api.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+export const getEndpointUrl = (
+    endpoint: string,
+    fusionEnv: FusionEnv,
+    version: string = '1.0-preview',
+) => {
+    const { FUSION_APP_API: appApiUrl } = process.env;
+    return (
+        appApiUrl ??
+        `https://fusion-s-apps-${fusionEnv}.azurewebsites.net/${endpoint}?api-version=${version}`
+    );
```
Comment (odinr) [inline packages/cli/src/bin/utils/app-api.ts:18]: i think we should resolve the app service from service discovery, but `preview` might not be there yet

File: packages/cli/src/bin/utils/app-api.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+};
+
+export const validateToken = () => {
+    const spinner = Spinner.Current;
+    spinner.info('Validating FUSION_TOKEN');
+
+    if (!process?.env?.FUSION_TOKEN) {
+        spinner.fail('😞', chalk.yellowBright('FUSION_TOKEN not found'));
```
Comment (odinr) [inline packages/cli/src/bin/utils/app-api.ts:26]: be more verbose, imagine debugging a workflow.

File: packages/cli/src/lib/app-config.ts [hunk 3]
```diff
@@ -11,12 +11,18 @@ import { AssertionError, assertObject } from './utils/assert.js';
 import { ResolvedAppPackage } from './app-package.js';
 import deepmerge from 'deepmerge/index.js';
 
+type Endpoints = {
```
Comment (odinr) [inline packages/cli/src/lib/app-config.ts:14]: think this should be `Record<name: string, {url: string, scopes?: string[]}`

it easier to access a record than filtering arrays

Comment (eikeland) [reply packages/cli/src/lib/app-config.ts:14]: i totally agree, but thats the format the API requires. Should i ask them to change it?

File: packages/cli/src/bin/utils/app-api.ts (hunk 2 above)
Comment (eikeland) [reply packages/cli/src/bin/utils/app-api.ts:26]: Done

File: packages/cli/src/bin/utils/app-api.ts (hunk 1 above)
Comment (eikeland) [reply packages/cli/src/bin/utils/app-api.ts]: Done

File: packages/cli/docs/api-authentication.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        run: pnpm install
+
+      - name: Pack, upload and tag application
+        shell: bash
+        run: pnpm exec fusion-framework-cli app publish --tag ${{inputs.tag}} --env ${{inputs.fusionEnv}}
+
+      - name: Upload config for this version and env
+        shell: bash
+        run: pnpm exec fusion-framework-cli app config --config app.config-ci.ts --publish ${{inputs.tag}} --env ${{inputs.fusionEnv}}
+
```
Comment (odinr) [inline packages/cli/docs/api-authentication.md:101]: remove pnpm from script, just use plain node and the install should just be conceptual

File: packages/cli/docs/commands.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+```sh
+  -p, --port <number>        dev-server port
+  -P, --portal <string>      fusion portal host
+  -c, --config <file>        use specified application config, by default search for app.config.{ts,js,json}
+      --manifest <file>      use specified manifest, by default search for app.manifest.config.{ts,js,json}
+      --vite <file>          use specified Vite config file, by default search for app.vite.config.{ts,js,json}
+  -F, --framework <string>   application framework to build the application on, supported: [react] (default: "react")
+  -d, --dev-portal <string>  Location of dev-portal you want to use
+  -h, --help                 display help for command
+```
```
Comment (odinr) [inline packages/cli/docs/commands.md:58]: change to markdown tables, repeate for all command options

File: packages/cli/docs/commands.md [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+  - name: Azure CLI
+    uses: azure/cli@v2
+    with:
+      azcliversion: latest
+      inlineScript: |
+        export FUSION_TOKEN="$(az account get-access-token --scope example-env-scope\.default)"
+  
+  - name: Publish app...
+```
```
Comment (odinr) [inline packages/cli/docs/commands.md:294]: move and and link to sample yml file

File: packages/react/framework/src/app/useMyApps.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import { ApplicationManifest } from '@equinor/fusion-framework-module-app';
```
Comment (odinr) [inline packages/react/framework/src/app/useMyApps.ts:1]: remove

File: packages/cli/docs/commands.md (hunk 6 above)
Comment (eikeland) [reply packages/cli/docs/commands.md:294]: Done

File: packages/cli/docs/configuration.md [hunk 8]
```diff
@@ -10,26 +10,50 @@ title: Fusion Framework CLI - App Configuration
...
-export type AppConfig = {
-    /** application config */
-    environment?: Record<string, unknown>;
+// app.config.ts
+import { mergeAppConfigs, defineAppConfig } from '@equinor/fusion-framework-cli';
+export default defineAppConfig((_env, { base }) =>
+    mergeAppConfigs(base, {
+        environment: {
+            endpoints: {
+              api: {
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/docs/configuration.md:22]: Døne

File: packages/cli/package.json [hunk 9]
```diff
@@ -41,11 +41,13 @@
         "@equinor/fusion-wc-chip": "^1.2.1",
         "@equinor/fusion-wc-person": "^3.0.1",
         "@types/adm-zip": "^0.5.0",
+        "@types/cors": "^2.8.17",
         "@types/semver": "^7.5.0",
         "@vitejs/plugin-react": "^4.0.4",
         "adm-zip": "^0.5.10",
         "chalk": "^5.3.0",
         "commander": "^12.0.0",
+        "cors": "^2.8.5",
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/package.json:50]: Yes, in dev-proxy.ts

File: packages/cli/src/bin/create-export-manifest.ts [hunk 10]
```diff
@@ -52,10 +45,14 @@ export const createExportManifest = async (options?: {
...
-        version: normalizeVersion(manifest.version),
-    };
+    // TODO - mutation is the mother of all f* ups
+    manifest.githubRepo ??= execSync('git remote get-url origin')
+        .toString()
+        .trim()
+        .replace('git@github.com:', 'https://github.com/')
+        .replace(/.git$/, '');
+    manifest.commitSha = execSync('git rev-parse HEAD').toString().trim();
+    manifest.timestamp = new Date().toISOString();
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/src/bin/create-export-manifest.ts:55]: Døne

File: packages/cli/src/bin/create-export-manifest.ts [hunk 11]
```diff
@@ -52,10 +45,14 @@ export const createExportManifest = async (options?: {
...
-        ...manifest,
-        version: normalizeVersion(manifest.version),
-    };
+    // TODO - mutation is the mother of all f* ups
+    manifest.githubRepo ??= execSync('git remote get-url origin')
+        .toString()
+        .trim()
+        .replace('git@github.com:', 'https://github.com/')
+        .replace(/.git$/, '');
+    manifest.commitSha = execSync('git rev-parse HEAD').toString().trim();
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/src/bin/create-export-manifest.ts:54]: Døne

File: packages/cli/src/bin/publish-application.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    });
+
+    spinner.info('Uploading bundle');
+    const uploadedBundle = await uploadAppBundle(appKey, 'app-bundle.zip', env, service);
+    if (!uploadedBundle) {
+        return;
+    }
+
+    spinner.info(`Tag app bundle with: ${tag}`);
+    const tagd = await tagAppBundle(tag, appKey, uploadedBundle.version, env, service);
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/src/bin/publish-application.ts:53]: Done

File: packages/cli/src/bin/publish-application.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+    if (!tagd) {
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/src/bin/publish-application.ts:55]: Done

File: packages/cli/src/bin/tag-application.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    spinner.info(`Tag app "${appKey}@${version}" with: "${tag}"`);
+
+    spinner.info('Verifying that App is registered');
+    const registered = await appRegistered(appKey, env, service);
+    if (!registered) {
+        return;
+    }
+
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/src/bin/tag-application.ts:42]: Done

File: packages/react/framework/src/app/index.ts [hunk 15]
```diff
@@ -4,4 +4,5 @@ export { useCurrentApp } from './useCurrentApp';
 export { useCurrentAppModule } from './useCurrentAppModule';
 export { useCurrentAppModules } from './useCurrentAppModules';
 export { useApps } from './useApps';
+export { useMyApps } from './useMyApps';
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/react/framework/src/app/index.ts:7]: No longer needed

File: packages/modules/app/src/app/App.ts [hunk 16]
```diff
@@ -445,13 +446,20 @@ export class App<TEnv = any, TModules extends Array<AnyModule> | unknown = unkno
...
         const manifest = await this.getManifestAsync(allow_cache);
-        this.#state.next(actions.importApp(manifest.entry));
+        if (!manifest.build.entryPoint) {
+            console.log(
+                `The ${manifest.key} is missing entryPoint, please upload a build for the app before continuing`,
+            );
+            return;
+        }
+
+        this.#state.next(actions.importApp(manifest.build.entryPoint));
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/modules/app/src/app/App.ts:462]: done

File: packages/modules/app/src/app/App.ts [hunk 17]
```diff
@@ -602,8 +610,14 @@ export class App<TEnv = any, TModules extends Array<AnyModule> | unknown = unkno
             subscriber.add(
                 // fetch application latest manifest and request loading of the application script
                 this.getManifest().subscribe((manifest) => {
+                    if (!manifest.build.entryPoint) {
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/modules/app/src/app/App.ts:613]: done

File: packages/modules/app/src/AppConfigurator.ts [hunk 18]
```diff
@@ -55,59 +54,99 @@ export class AppConfigurator implements IAppConfigurator {
 
             // TODO - remove when refactor portal service!
             /** resolve and create a client from discovery */
-            try {
-                return await serviceDiscovery.createClient('app');
-            } catch {
-                return await serviceDiscovery.createClient('portal');
-            }
+            return await serviceDiscovery.createClient('apps', {
+                onCreate(client) {
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/modules/app/src/AppConfigurator.ts:60]: Done

File: packages/modules/app/src/enable-app-module.ts [hunk 19]
```diff
@@ -10,12 +9,14 @@ import { module } from './module';
 export const enableAppModule = (
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     configurator: IModulesConfigurator<any, any>,
-    builder?: AppConfigBuilderCallback,
+    callback?: (builder: AppConfigurator) => void,
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/modules/app/src/enable-app-module.ts:12]: Done

File: packages/modules/app/src/AppConfigurator.ts [hunk 20]
```diff
@@ -1,44 +1,43 @@
...
 export interface AppModuleConfig {
-    proxy: {
-        path: string;
-    },
     client: {
-        getAppManifest: QueryCtorOptions<AppManifest, { appKey: string }>;
-        getAppManifests: QueryCtorOptions<AppManifest[], void>;
+        getAppManifest: QueryCtorOptions<ApplicationManifest, { appKey: string }>;
+        getAppManifests: QueryCtorOptions<ApplicationManifest[], void>;
+        getMyAppManifests: QueryCtorOptions<ApplicationManifest[], void>;
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/modules/app/src/AppConfigurator.ts]: Never been used before so no breaking change

File: packages/react/framework/src/app/useMyApps.ts (hunk 7 above)
Comment (eikeland) [reply packages/react/framework/src/app/useMyApps.ts]: done

File: packages/cli/src/bin/tag-application.ts [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+    env: FusionEnv;
+    service: string;
+}) => {
+    const { tag, version, env, service } = options;
+
+    const spinner = Spinner.Global({ prefixText: chalk.dim('Tag') });
+
+    const validToken = validateToken();
+    if (!validToken) {
+        return;
```
Comment (odinr) [inline packages/cli/src/bin/tag-application.ts:23]: instead of return, throw error

File: packages/cli/src/bin/tag-application.ts [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+    spinner.info(`Tag app "${appKey}@${version}" with: "${tag}"`);
+
+    spinner.info('Verifying that App is registered');
+    const registered = await appRegistered(appKey, env, service);
+    if (!registered) {
+        return;
+    }
+
+    const tagged = await tagAppBundle(tag, appKey, version, env, service);
```
Comment (odinr) [inline packages/cli/src/bin/tag-application.ts:43]: wrap with try catch, no need to 
```suggestion
try {
    await tagAppBundle(tag, appKey, version, env, service);
    spinner.succeed(
        '✅',
        `App: "${chalk.greenBright(appKey)}"`,
        `Version: "${chalk.greenBright(tagged.version)}"`,
        `Tag: "${chalk.greenBright(tagged.tagName)}"`,
    );
} catch(err) {
  spinner.failed(....)
  throw Error('booo')
}
```

File: packages/cli/src/lib/app-manifest.ts [hunk 23]
```diff
@@ -127,33 +119,58 @@ export const resolveManifest = async (
...
+                .trim()
+                .replace('git@github.com:', 'https://github.com/')
+                .replace(/.git$/, '');
+        }
+    } catch {
+        return undefined;
+    }
+};
+
+const getGitCommitSha = () => {
```
Comment (odinr) [inline packages/cli/src/lib/app-manifest.ts:140]:
```suggestion
const resolveGitCommitSha = () => {
```

File: packages/cli/src/bin/tag-application.ts (hunk 22 above)
Comment (eikeland) [reply packages/cli/src/bin/tag-application.ts:43]: Done

File: packages/cli/src/bin/tag-application.ts (hunk 21 above)
Comment (eikeland) [reply packages/cli/src/bin/tag-application.ts:23]: i would like to use exit(1) here. maybe we can adda flag for verbose later on and throw all error then?

File: packages/cli/src/lib/app-manifest.ts [hunk 24]
```diff
@@ -127,33 +119,58 @@ export const resolveManifest = async (
     return resolveConfig(manifestConfigFilename, { find: options });
 };
 
-export const createManifestFromPackage = (pkg: ResolvedAppPackage): AppManifest => {
+const getGithubRepo = (pkg: AppPackageJson) => {
```
[the remark this answers is not in the document]
Comment (eikeland) [reply packages/cli/src/lib/app-manifest.ts:122]: Done

File: packages/cli/src/lib/app-manifest.ts (hunk 23 above)
Comment (eikeland) [reply packages/cli/src/lib/app-manifest.ts]: Done

**Comment by odinr** [comment]
@eikeland missing changeset for `docs` (vue-press)

File: packages/cli/src/bin/utils/getEndpointUrl.ts [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+    if (service || CUSTOM_APPAPI) {
+        return service ?? CUSTOM_APPAPI;
+    }
+
+    /* Env has changed get new api url */
+    if (FUSION_CLI_ENV !== fusionEnv || !process.env.FUSION_CLI_APPAPI) {
+        process.env.FUSION_CLI_ENV = fusionEnv;
+
+        const requestService = await fetch(
+            `https://discovery.ci.fusion-dev.net/service-registry/environments/${fusionEnv}/services/apps`,
```
Comment (odinr) [inline packages/cli/src/bin/utils/getEndpointUrl.ts:30]: This should use the public discovery url

File: packages/cli/src/bin/utils/publishAppConfig.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+        );
+    } else if (!requestConfig.ok || requestConfig.status > 399) {
+        const response = await requestConfig.json();
+        console.log(response);
+        throw new Error(
+            `Failed to upload config. HTTP status ${requestConfig.status}, ${requestConfig.statusText}`,
+        );
+    }
+
+    return await requestConfig.json();
```
Comment (odinr) [inline packages/cli/src/bin/utils/publishAppConfig.ts:32]: Unnecessary await keyword, but this should be wrapped in try catch, case invalid response.

File: packages/cli/src/bin/utils/tagAppBundle.ts [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+
+    if (requestTag.status !== 200) {
+        const response = await requestTag.json();
+        console.log(response);
+        throw new Error(
+            `Failed to tag bundle. HTTP status ${requestTag.status}, ${requestTag.statusText}`,
+        );
+    }
+
+    return await requestTag.json();
```
Comment (odinr) [inline packages/cli/src/bin/utils/tagAppBundle.ts:31]: Unnecessary keyword

File: packages/cli/src/bin/utils/uploadAppBundle.ts [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+
+    if (!requestBundle.ok) {
+        const json = await requestBundle.json();
+        console.error(json);
+        throw new Error(
+            `Failed to publish bundle. HTTP status ${requestBundle.status}, ${requestBundle.statusText}`,
+        );
+    }
+
+    return await requestBundle.json();
```
Comment (odinr) [inline packages/cli/src/bin/utils/uploadAppBundle.ts:57]: Unnecessary keyword

File: packages/cli/src/bin/utils/getEndpointUrl.ts (hunk 25 above)
Comment (eikeland) [reply packages/cli/src/bin/utils/getEndpointUrl.ts:30]: Done....

File: packages/cli/src/bin/utils/tagAppBundle.ts (hunk 27 above)
Comment (eikeland) [reply packages/cli/src/bin/utils/tagAppBundle.ts:31]: Done

File: packages/cli/src/bin/utils/uploadAppBundle.ts (hunk 28 above)
Comment (eikeland) [reply packages/cli/src/bin/utils/uploadAppBundle.ts:57]: Done

**Pull request merged**